Source/NameServer/AddressBook/AddressBook.cpp

```diff
@@ -0,0 +1,1367 @@
+{
+    for(std::vector<DevTypeRecord_t>::const_iterator D=Data.DeviceTypes.begin();
+            D!=Data.DeviceTypes.end(); D++)
+    {
+        if (unsigned retVal = ValidateDeviceType(*D, Data.MessageTypes.size()))
+            return retVal;
+    }
+
+    return SUCCESS;
+}
+
+/*==============================================================================
+ * ValidateDeviceType:  Check that DeviceType Message indices are valid and not 
+ *                  out of range.
+ * Added and separated from ValidateTask 5 July 2019 ADR
+ *============================================================================*/
+unsigned AddressBook::ValidateDeviceType(const DevTypeRecord_t& DevTypRec, unsigned MaxIdx)
+{
+    for(std::vector<MsgIdx>::const_iterator I=DevTypRec.InMsgs.begin();
+        I!=DevTypRec.InMsgs.end(); I++)
+    {
+        if(*I >= MaxIdx)
+        {
+            return(ERR_INVALID_MESSAGE_TYPE);
+        }
+    }
+
+    // changed to iterate over OuMsgs (was InMsgs) 5 July 2019 ADR
+    for(std::vector<MsgIdx>::const_iterator O=DevTypRec.OuMsgs.begin();
+        O!=DevTypRec.OuMsgs.end(); O++)
+    {
+        if(*O >= MaxIdx)
+        {
+            return(ERR_INVALID_MESSAGE_TYPE);
+        }
+    }
+    return 0;
+}
+
+
+/*==============================================================================
+ * ValidateDevice:  Check that the device does not already exist in the name map
+ *                  and that indices in the Record are not out of range.
+ *============================================================================*/
+unsigned AddressBook::ValidateDevice(TaskRecord_t *TRec, Record_t &DevRec)
+{
+    // Check that the device does not already exist in the maps
+    AddrMap_t::const_iterator DASearch = TRec->AddrMap.find(DevRec.Address);
+    if (DASearch != TRec->AddrMap.end()) {
+        // DebugPrint("Device address %llu already exists\n", DevRec.Address);
+        return(ERR_DEVICE_ADDR_USED);
+    }
+
+    NameMap_t::const_iterator DNSearch = TRec->NameMap.find(DevRec.Name);
+    if (DNSearch != TRec->NameMap.end()) {
+        // DebugPrint("Device name %s already exists\n", DevRec.Name.c_str());
+        return(ERR_DEVICENAME_USED);
+    }
+
+    // Check that the indices are valid BEFORE adding the Device.
+    if (DevRec.DeviceType >= TRec->DevTypes.size()) {
+        // DebugPrint("Device type %u does not exist\n", DevRec.DeviceType);
+        return(ERR_INVALID_DEVTYPE);
+    }
+
+    if (DevRec.Attribute >= static_cast<int>(TRec->AttrTypes.size())) {
+        // DebugPrint("Device attribute type %u does not exist\n", DevRec.Attribute);
+        return(ERR_INVALID_ATTRIBUTE);
+    }
+
+    return SUCCESS;   // Validation Successful
+}
+
+/*==============================================================================
+ * CheckVector: Check that adding a Record to the vector won't cause a
+ *              reallocation as this would silently invalidate all of the maps.
```

Comment:
Why would reallocation invalidate all of the maps? Is it because the vector holds elements (as opposed to pointers to elements), where the maps hold pointers to those elements?

Reply:
Exactly